internal/ethapi/api.go

```diff
@@ -790,18 +790,32 @@ func (s *PublicBlockChainAPI) EstimateGas(ctx context.Context, args CallArgs) (h
 	// Execute the binary search and hone in on an executable gas limit
 	for lo+1 < hi {
 		mid := (hi + lo) / 2
-		if !executable(mid) {
+		if !executable(mid) { // Assume failure from insufficient gas
 			lo = mid
 		} else {
 			hi = mid
 		}
 	}
-	// Reject the transaction as invalid if it still fails at the highest allowance
-	if hi == cap {
-		if !executable(hi) {
-			return 0, fmt.Errorf("gas required exceeds allowance or always failing transaction")
+	// If transaction fails at highest allowance, check if lower gas executes
+	// as failure could be due to insufficient balance to pay gas
+	if hi == cap && !executable(hi) {
+		lo = params.TxGas - 1 // Re initialize to lower limit
+		for lo + 1 < hi {
+			mid := (hi + lo) / 2
+			if !executable(mid) { // Assume failure from insufficent balance
+				hi = mid
+			} else {
+				lo = mid
+			}
 		}
+		hi = lo // Hi is one too high to execute, set to executable value
```

Comment:
Not sure what geth storage costs are like, can make another var `gas` if cheap and set `gas = hi` in case above, `gas = lo` here to make more readable

Reply:
You don't need to worry about EVM storage costs here, if that's what you are talking about. This code is only run as an RPC to request a fee estimate, not as part of the consensus.